From 185ddfd1d56eb6ed4b7470d1713a2270a2d3bedf Mon Sep 17 00:00:00 2001
From: Alison Chaiken <alison_chaiken@mentor.com>
Date: Thu, 16 Jul 2015 09:38:55 +0200
Subject: [PATCH 07/26] gstwaylandsink: fix OMX-compatible caps for Venus

Force waylandsink to support the NV12 format required by omxh264dec.
Without further work, a decoder pipeline with

. . . omxh264dec ! waylandsink

will link, but then will hang.

The alternative of setting a proprietary image format
explicitly on the decoder source pad won't work, as
gst_omx_video_get_supported_colorformats ()
will report that the decoder HW doesn't support it.

Signed-off-by: Alison Chaiken <alison_chaiken@mentor.com>
---
 ext/wayland/gstwaylandsink.c | 50 ++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 48 insertions(+), 2 deletions(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index d029d83..5866051 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -65,13 +65,21 @@ enum
 GST_DEBUG_CATEGORY (gstwayland_debug);
 #define GST_CAT_DEFAULT gstwayland_debug
 
+#define USE_OMX_TARGET_VENUS
+
 static GstStaticPadTemplate sink_template = GST_STATIC_PAD_TEMPLATE ("sink",
     GST_PAD_SINK,
     GST_PAD_ALWAYS,
+#if (defined (USE_OMX_TARGET_VENUS))
+    GST_STATIC_CAPS ("video/x-raw, format = NV12,"
+      "width = 1920, "
+      "height = 1080, framerate = 25/1")
+#else
     GST_STATIC_CAPS (GST_VIDEO_CAPS_MAKE
         ("{ BGRx, BGRA, RGBx, xBGR, xRGB, RGBA, ABGR, ARGB, RGB, BGR, "
             "RGB16, BGR16, YUY2, YVYU, UYVY, AYUV, NV12, NV21, NV16, "
             "YUV9, YVU9, Y41B, I420, YV12, Y42B, v308 }"))
+#endif
     );
 
 static void gst_wayland_sink_get_property (GObject * object,
@@ -411,7 +419,7 @@ static GstCaps *
 gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
 {
   GstWaylandSink *sink;
-  GstCaps *caps;
+  GstCaps *caps, *temp;
 
   sink = GST_WAYLAND_SINK (bsink);
 
@@ -424,21 +432,51 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
     GValue value = G_VALUE_INIT;
     GArray *formats;
     gint i;
-    enum wl_shm_format fmt;
+    enum wl_shm_format fmt, format;
+    GstVideoInfo info;
 
     g_value_init (&list, GST_TYPE_LIST);
     g_value_init (&value, G_TYPE_STRING);
 
+    /* No negotiation for size or fps: fix later */
+#if (defined (USE_OMX_TARGET_VENUS))
+    GST_DEBUG_OBJECT (sink, "Caps for Venus: %" GST_PTR_FORMAT, caps);
+
+    /* extract info from caps */
+    if (!gst_video_info_from_caps (&info, caps)) {
+      GST_ERROR_OBJECT(sink, "No video_info in caps.");
+      g_mutex_unlock (&sink->display_lock);
+      return NULL;
+    }
+
+    format = gst_video_format_to_wayland_format (GST_VIDEO_INFO_FORMAT (&info));
+   if ((gint) format == -1) {
+      GST_ERROR_OBJECT(sink, "Wayland format conversion failed.");
+      g_mutex_unlock (&sink->display_lock);
+      return NULL;
+    }
+
+   g_value_set_string (&value, gst_wayland_format_to_string (format));
+   /* list was empty */
+   gst_value_list_append_value (&list, &value);
+
+#else
     formats = sink->display->formats;
     for (i = 0; i < formats->len; i++) {
       fmt = g_array_index (formats, uint32_t, i);
       g_value_set_string (&value, gst_wayland_format_to_string (fmt));
       gst_value_list_append_value (&list, &value);
     }
+#endif
 
     caps = gst_caps_make_writable (caps);
     gst_structure_set_value (gst_caps_get_structure (caps, 0), "format", &list);
 
+#if defined (USE_OMX_TARGET_VENUS)
+    temp = caps;
+    caps = gst_caps_truncate(temp);
+#endif
+
     GST_DEBUG_OBJECT (sink, "display caps: %" GST_PTR_FORMAT, caps);
   }
 
@@ -480,6 +518,14 @@ gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
   if ((gint) format == -1)
     goto invalid_format;
 
+#if (defined (USE_OMX_TARGET_VENUS))
+  /* store the video info */
+  sink->video_info = info;
+  sink->video_info_changed = TRUE;
+
+  return TRUE;
+#endif
+
   /* verify we support the requested format */
   formats = sink->display->formats;
   for (i = 0; i < formats->len; i++) {
-- 
1.9.1

